File: EscapeNine-endless-/EscapeNine-endless-/Models/Character.swift
//
//  Character.swift
//  EscapeNine-endless-
//
//  Created by 吉留徹 on 2025/11/14.
//

import Foundation

enum CharacterType: String, CaseIterable {
    case hero = "hero"
    case thief = "thief"
    case wizard = "wizard"
    case elf = "elf"
    case knight = "knight"

    var name: String {
        switch self {
        case .hero: return "勇者"
        case .thief: return "盗賊"
        case .wizard: return "魔法使い"
        case .elf: return "エルフ"
        case .knight: return "ナイト"
        }
    }

    var isFree: Bool {
        switch self {
        case .hero, .thief: return true
        case .wizard, .elf, .knight: return false
        }
    }

    var price: Int? {
        switch self {
        case .hero, .thief: return nil
        case .wizard, .elf, .knight: return Constants.premiumCharacterPrice
        }
    }
}

struct Character {
    let id: String
    let type: CharacterType
    let name: String
    let skill: Skill
    let isFree: Bool
    let price: Int?
    let spriteName: String
    
    static func getCharacter(for type: CharacterType) -> Character {
        let skill: Skill
        switch type {
        case .hero:
            skill = Skill(type: .dash, name: "ダッシュ", description: "2マス移動できる", maxUsage: Constants.heroSkillMaxUsage)
        case .thief:
            skill = Skill(type: .diagonal, name: "斜め移動", description: "斜め方向に移動可能", maxUsage: Constants.thiefSkillMaxUsage)
        case .wizard:
            skill = Skill(type: .invisible, name: "透明化", description: "鬼に当たっても無敵", maxUsage: Constants.wizardSkillMaxUsage)
        case .elf:
            skill = Skill(type: .bind, name: "拘束", description: "鬼を\(Constants.bindDurationTurns)ターン停止させる", maxUsage: Constants.elfSkillMaxUsage)
        case .knight:
            skill = Skill(type: .shield, name: "盾ガード", description: "次の衝突を1回無効化する", maxUsage: Constants.knightSkillMaxUsage)
        }

        return Character(
            id: type.rawValue,
            type: type,
            name: type.name,
            skill: skill,
            isFree: type.isFree,
            price: type.price,
            spriteName: type.rawValue
        )
    }
    
    static var defaultCharacter: Character {
        return getCharacter(for: .hero)
    }
}

File: unity/EscapeNine/Assets/Scripts/Core/Character.cs
// Character.cs
// Swift 正本からの忠実移植: Models/Character.swift

namespace EscapeNine.Core
{
    public static class CharacterTypeInfo
    {
        public static string Name(this CharacterType type) => type switch
        {
            CharacterType.Hero => "勇者",
            CharacterType.Thief => "盗賊",
            CharacterType.Wizard => "魔法使い",
            CharacterType.Elf => "エルフ",
            CharacterType.Knight => "ナイト",
            _ => ""
        };

        public static bool IsFree(this CharacterType type) => type switch
        {
            CharacterType.Hero => true,
            CharacterType.Thief => true,
            _ => false // wizard / elf / knight は有料
        };

        /// <summary>価格 (無料キャラは null)。Swift: CharacterType.price</summary>
        public static int? Price(this CharacterType type) =>
            type.IsFree() ? (int?)null : GameConfig.PremiumCharacterPrice;
    }

    /// <summary>キャラクター定義。Swift: struct Character</summary>
    public readonly struct Character
    {
        public readonly string Id;
        public readonly CharacterType Type;
        public readonly string Name;
        public readonly Skill Skill;
        public readonly bool IsFree;
        public readonly int? Price;
        public readonly string SpriteName;

        public Character(string id, CharacterType type, string name, Skill skill, bool isFree, int? price, string spriteName)
        {
            Id = id;
            Type = type;
            Name = name;
            Skill = skill;
            IsFree = isFree;
            Price = price;
            SpriteName = spriteName;
        }

        /// <summary>種別からキャラを生成。Swift: Character.getCharacter(for:)</summary>
        public static Character GetCharacter(CharacterType type)
        {
            Skill skill;
            switch (type)
            {
                case CharacterType.Hero:
                    skill = new Skill(SkillType.Dash, "ダッシュ", "2マス移動できる", GameConfig.HeroSkillMaxUsage);
                    break;
                case CharacterType.Thief:
                    skill = new Skill(SkillType.Diagonal, "斜め移動", "斜め方向に移動可能", GameConfig.ThiefSkillMaxUsage);
                    break;
                case CharacterType.Wizard:
                    skill = new Skill(SkillType.Invisible, "透明化", "鬼に当たっても無敵", GameConfig.WizardSkillMaxUsage);
                    break;
                case CharacterType.Elf:
                    skill = new Skill(SkillType.Bind, "拘束", $"鬼を{GameConfig.BindDurationTurns}ターン停止させる", GameConfig.ElfSkillMaxUsage);
                    break;
                case CharacterType.Knight:
                    skill = new Skill(SkillType.Shield, "盾ガード", "次の衝突を1回無効化する", GameConfig.KnightSkillMaxUsage);
                    break;
                default:
                    skill = new Skill(SkillType.Dash, "ダッシュ", "2マス移動できる", GameConfig.HeroSkillMaxUsage);
                    break;
            }

            return new Character(
                id: type.RawValue(),
                type: type,
                name: type.Name(),
                skill: skill,
                isFree: type.IsFree(),
                price: type.Price(),
                spriteName: type.RawValue()
            );
        }

        public static Character DefaultCharacter => GetCharacter(CharacterType.Hero);
    }
}
